### PR TITLE
allow LSP insert text to replace non-matching prefixes (#5469)

### LINKED ISSUE #5469: Pyright LSP completion lowercase to uppercase leaves first few characters
### Summary

When I use Pyright completion that changes case from lowercase to uppercase, the first two characters remain instead of being replaced.

The same does not happen using the same language server in neovim, so I don't think it's an issue with the language server itself.

### Reproduction Steps

Using this `languages.toml` entry for Pyright:

```toml
[[language]]
name = "python"

[language.language-server]
command = "pyright-langserver"
args = ["--stdio"]

[language.config.python.analysis]
autoSearchPaths = true
diagnosticMode = "workspace"
useLibraryCodeForTypes = true
```

If I create a simple python project with a virtualenvironment and activate the language server:

```sh
$ cd $(mktemp -d)
$ touch pyproject.toml
$ mkdir src
$ python -mvenv ./venv
$ . ./venv/bin/activate
$ pip install --force pyright
$ hx src/example.py
```
Then add a simple uppercase `Uppercase` class, and start typing `uppercase` and complete it to the class, I end up with this:

```py
class Uppercase:
    pass

upUppercase
```

If I type it in the right casing to begin with, it has no problem and does not duplicate the first two letters.

### Helix log

```
2023-01-09T10:07:44.240 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"textDocument/didChange","params":{"contentChanges":[{"range":{"end":{"character":0,"line":2},"start":{"character":0,"line":2}},"text":"\n"}],"textDocument":{"uri":"file:///tmp/tmp.ZmHxbsplID/src/example.py","version":23}}}
2023-01-09T10:07:44.501 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///tmp/tmp.ZmHxbsplID/src/example.py","version":23,"diagnostics":[]}}
2023-01-09T10:07:44.800 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"textDocument/didChange","params":{"contentChanges":[{"range":{"end":{"character":0,"line":3},"start":{"character":0,"line":3}},"text":"u"}],"textDocument":{"uri":"file:///tmp/tmp.ZmHxbsplID/src/example.py","version":24}}}
2023-01-09T10:07:44.984 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"textDocument/didChange","params":{"contentChanges":[{"range":{"end":{"character":1,"line":3},"start":{"character":1,"line":3}},"text":"p"}],"textDocument":{"uri":"file:///tmp/tmp.ZmHxbsplID/src/example.py","version":25}}}
2023-01-09T10:07:44.991 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"textDocument/completion","params":{"position":{"character":2,"line":3},"textDocument":{"uri":"file:///tmp/tmp.ZmHxbsplID/src/example.py"}},"id":12}
2023-01-09T10:07:44.994 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","id":12,"result":{"items":[{"label":"Uppercase","kind":7,"data":{"workspacePath":"/tmp/tmp.ZmHxbsplID","filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"symbolLabel":"Uppercase"},"sortText":"09.9999.Uppercase"},{"label":"super","kind":7,"data":{"workspacePath":"/tmp/tmp.ZmHxbsplID","filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"symbolLabel":"super"},"sortText":"09.9999.super"},{"label":"tuple","kind":7,"data":{"workspacePath":"/tmp/tmp.ZmHxbsplID","filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"symbolLabel":"tuple"},"sortText":"09.9999.tuple"},{"label":"KeyboardInterrupt","kind":7,"data":{"workspacePath":"/tmp/tmp.ZmHxbsplID","filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"symbolLabel":"KeyboardInterrupt"},"sortText":"09.9999.KeyboardInterrupt"},{"label":"LookupError","kind":7,"data":{"workspacePath":"/tmp/tmp.ZmHxbsplID","filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"symbolLabel":"LookupError"},"sortText":"09.9999.LookupError"},{"label":"InterruptedError","kind":7,"data":{"workspacePath":"/tmp/tmp.ZmHxbsplID","filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"symbolLabel":"InterruptedError"},"sortText":"09.9999.InterruptedError"},{"label":"ProcessLookupError","kind":7,"data":{"workspacePath":"/tmp/tmp.ZmHxbsplID","filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"symbolLabel":"ProcessLookupError"},"sortText":"09.9999.ProcessLookupError"},{"label":"BaseExceptionGroup","kind":7,"data":{"workspacePath":"/tmp/tmp.ZmHxbsplID","filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"symbolLabel":"BaseExceptionGroup"},"sortText":"09.9999.BaseExceptionGroup"},{"label":"ExceptionGroup","kind":7,"data":{"workspacePath":"/tmp/tmp.ZmHxbsplID","filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"symbolLabel":"ExceptionGroup"},"sortText":"09.9999.ExceptionGroup"},{"label":"UnionType","kind":7,"data":{"workspacePath":"/tmp/tmp.ZmHxbsplID","filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"autoImportText":"from types import UnionType","symbolLabel":"UnionType"},"sortText":"12.9999.UnionType","detail":"Auto-import","labelDetails":{"description":"types"},"documentation":{"kind":"plaintext","value":"from types import UnionType"},"textEdit":{"range":{"start":{"line":3,"character":0},"end":{"line":3,"character":2}},"newText":"UnionType"}},{"label":"Unpack","kind":7,"data":{"workspacePath":"/tmp/tmp.ZmHxbsplID","filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"autoImportText":"from typing import Unpack","symbolLabel":"Unpack"},"sortText":"12.9999.Unpack","detail":"Auto-import","labelDetails":{"description":"typing"},"documentation":{"kind":"plaintext","value":"from typing import Unpack"},"textEdit":{"range":{"start":{"line":3,"character":0},"end":{"line":3,"character":2}},"newText":"Unpack"}},{"label":"Unpack","kind":7,"data":{"workspacePath":"/tmp/tmp.ZmHxbsplID","filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"autoImportText":"from typing_extensions import Unpack","symbolLabel":"Unpack"},"sortText":"12.9999.Unpack","detail":"Auto-import","labelDetails":{"description":"typing_extensions"},"documentation":{"kind":"plaintext","value":"from typing_extensions import Unpack"},"textEdit":{"range":{"start":{"line":3,"character":0},"end":{"line":3,"character":2}},"newText":"Unpack"}},{"label":"update_abstractmethods","kind":3,"data":{"workspacePath":"/tmp/tmp.ZmHxbsplID","filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"autoImportText":"from abc import update_abstractmethods","symbolLabel":"update_abstractmethods"},"sortText":"12.9999.update_abstractmethods","detail":"Auto-import","labelDetails":{"description":"abc"},"documentation":{"kind":"plaintext","value":"from abc import update_abstractmethods"},"textEdit":{"range":{"start":{"line":3,"character":0},"end":{"line":3,"character":2}},"newText":"update_abstractmethods"}}],"isIncomplete":true}}
2023-01-09T10:07:44.994 helix_lsp::transport [INFO] <- {"isIncomplete":true,"items":[{"data":{"filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"symbolLabel":"Uppercase","workspacePath":"/tmp/tmp.ZmHxbsplID"},"kind":7,"label":"Uppercase","sortText":"09.9999.Uppercase"},{"data":{"filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"symbolLabel":"super","workspacePath":"/tmp/tmp.ZmHxbsplID"},"kind":7,"label":"super","sortText":"09.9999.super"},{"data":{"filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"symbolLabel":"tuple","workspacePath":"/tmp/tmp.ZmHxbsplID"},"kind":7,"label":"tuple","sortText":"09.9999.tuple"},{"data":{"filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"symbolLabel":"KeyboardInterrupt","workspacePath":"/tmp/tmp.ZmHxbsplID"},"kind":7,"label":"KeyboardInterrupt","sortText":"09.9999.KeyboardInterrupt"},{"data":{"filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"symbolLabel":"LookupError","workspacePath":"/tmp/tmp.ZmHxbsplID"},"kind":7,"label":"LookupError","sortText":"09.9999.LookupError"},{"data":{"filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"symbolLabel":"InterruptedError","workspacePath":"/tmp/tmp.ZmHxbsplID"},"kind":7,"label":"InterruptedError","sortText":"09.9999.InterruptedError"},{"data":{"filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"symbolLabel":"ProcessLookupError","workspacePath":"/tmp/tmp.ZmHxbsplID"},"kind":7,"label":"ProcessLookupError","sortText":"09.9999.ProcessLookupError"},{"data":{"filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"symbolLabel":"BaseExceptionGroup","workspacePath":"/tmp/tmp.ZmHxbsplID"},"kind":7,"label":"BaseExceptionGroup","sortText":"09.9999.BaseExceptionGroup"},{"data":{"filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"symbolLabel":"ExceptionGroup","workspacePath":"/tmp/tmp.ZmHxbsplID"},"kind":7,"label":"ExceptionGroup","sortText":"09.9999.ExceptionGroup"},{"data":{"autoImportText":"from types import UnionType","filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"symbolLabel":"UnionType","workspacePath":"/tmp/tmp.ZmHxbsplID"},"detail":"Auto-import","documentation":{"kind":"plaintext","value":"from types import UnionType"},"kind":7,"label":"UnionType","labelDetails":{"description":"types"},"sortText":"12.9999.UnionType","textEdit":{"newText":"UnionType","range":{"end":{"character":2,"line":3},"start":{"character":0,"line":3}}}},{"data":{"autoImportText":"from typing import Unpack","filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"symbolLabel":"Unpack","workspacePath":"/tmp/tmp.ZmHxbsplID"},"detail":"Auto-import","documentation":{"kind":"plaintext","value":"from typing import Unpack"},"kind":7,"label":"Unpack","labelDetails":{"description":"typing"},"sortText":"12.9999.Unpack","textEdit":{"newText":"Unpack","range":{"end":{"character":2,"line":3},"start":{"character":0,"line":3}}}},{"data":{"autoImportText":"from typing_extensions import Unpack","filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"symbolLabel":"Unpack","workspacePath":"/tmp/tmp.ZmHxbsplID"},"detail":"Auto-import","documentation":{"kind":"plaintext","value":"from typing_extensions import Unpack"},"kind":7,"label":"Unpack","labelDetails":{"description":"typing_extensions"},"sortText":"12.9999.Unpack","textEdit":{"newText":"Unpack","range":{"end":{"character":2,"line":3},"start":{"character":0,"line":3}}}},{"data":{"autoImportText":"from abc import update_abstractmethods","filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"symbolLabel":"update_abstractmethods","workspacePath":"/tmp/tmp.ZmHxbsplID"},"detail":"Auto-import","documentation":{"kind":"plaintext","value":"from abc import update_abstractmethods"},"kind":3,"label":"update_abstractmethods","labelDetails":{"description":"abc"},"sortText":"12.9999.update_abstractmethods","textEdit":{"newText":"update_abstractmethods","range":{"end":{"character":2,"line":3},"start":{"character":0,"line":3}}}}]}
2023-01-09T10:07:45.136 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"textDocument/didChange","params":{"contentChanges":[{"range":{"end":{"character":2,"line":3},"start":{"character":2,"line":3}},"text":"p"}],"textDocument":{"uri":"file:///tmp/tmp.ZmHxbsplID/src/example.py","version":26}}}
2023-01-09T10:07:45.352 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"textDocument/didChange","params":{"contentChanges":[{"range":{"end":{"character":3,"line":3},"start":{"character":3,"line":3}},"text":"e"}],"textDocument":{"uri":"file:///tmp/tmp.ZmHxbsplID/src/example.py","version":27}}}
2023-01-09T10:07:45.472 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"textDocument/didChange","params":{"contentChanges":[{"range":{"end":{"character":4,"line":3},"start":{"character":4,"line":3}},"text":"r"}],"textDocument":{"uri":"file:///tmp/tmp.ZmHxbsplID/src/example.py","version":28}}}
2023-01-09T10:07:45.730 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///tmp/tmp.ZmHxbsplID/src/example.py","version":28,"diagnostics":[{"range":{"start":{"line":3,"character":0},"end":{"line":3,"character":5}},"message":"\"upper\" is not defined","severity":1,"code":"reportUndefinedVariable","source":"Pyright","codeDescription":{"href":"https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportUndefinedVariable"}},{"range":{"start":{"line":3,"character":0},"end":{"line":3,"character":5}},"message":"Expression value is unused","severity":2,"code":"reportUnusedExpression","source":"Pyright","codeDescription":{"href":"https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportUnusedExpression"}}]}}
2023-01-09T10:07:45.992 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"textDocument/didChange","params":{"contentChanges":[{"range":{"end":{"character":5,"line":3},"start":{"character":2,"line":3}},"text":""}],"textDocument":{"uri":"file:///tmp/tmp.ZmHxbsplID/src/example.py","version":29}}}
2023-01-09T10:07:45.992 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"textDocument/didChange","params":{"contentChanges":[{"range":{"end":{"character":2,"line":3},"start":{"character":2,"line":3}},"text":"Uppercase"}],"textDocument":{"uri":"file:///tmp/tmp.ZmHxbsplID/src/example.py","version":30}}}
2023-01-09T10:07:45.999 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"completionItem/resolve","params":{"data":{"filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"symbolLabel":"Uppercase","workspacePath":"/tmp/tmp.ZmHxbsplID"},"kind":7,"label":"Uppercase","sortText":"09.9999.Uppercase"},"id":13}
2023-01-09T10:07:46.002 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","id":13,"result":{"data":{"filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"symbolLabel":"Uppercase","workspacePath":"/tmp/tmp.ZmHxbsplID"},"kind":7,"label":"Uppercase","sortText":"09.9999.Uppercase","documentation":{"kind":"plaintext","value":"class Uppercase()"}}}
2023-01-09T10:07:46.002 helix_lsp::transport [INFO] <- {"data":{"filePath":"/tmp/tmp.ZmHxbsplID/src/example.py","position":{"character":2,"line":3},"symbolLabel":"Uppercase","workspacePath":"/tmp/tmp.ZmHxbsplID"},"documentation":{"kind":"plaintext","value":"class Uppercase()"},"kind":7,"label":"Uppercase","sortText":"09.9999.Uppercase"}
2023-01-09T10:07:46.250 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///tmp/tmp.ZmHxbsplID/src/example.py","version":30,"diagnostics":[{"range":{"start":{"line":3,"character":0},"end":{"line":3,"character":11}},"message":"\"upUppercase\" is not defined","severity":1,"code":"reportUndefinedVariable","source":"Pyright","codeDescription":{"href":"https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportUndefinedVariable"}},{"range":{"start":{"line":3,"character":0},"end":{"line":3,"character":11}},"message":"Expression value is unused","severity":2,"code":"reportUnusedExpression","source":"Pyright","codeDescription":{"href":"https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportUnusedExpression"}}]}}

```

### Platform

Fedora Linux 37

### Terminal Emulator

gnome-terminal

### Helix Version

helix 22.12

### PR DESCRIPTION
**Pull request by Taywee**
closes #5469 

Most LSPs will complete case-insensitive matches, particularly from lowercase to uppercase.  In some cases, notably Pyright, this is given as a simple insert text instead of TextEdit.  When this happens, the prefix text was left unedited.

This change seems to fix it for me.  I'm not sure if there are other side effects to doing it this way, though.  If another LSP gave an insert text that was just the rest to be inserted, I imagine this would still cause problems.  But the previous method would also fail in that case with repeated text.  If you typed `su` and the suggestion was `surration`, it would be ambiguous in that case whether that was to be inserted (making `susurration`) or replaced entirely (making `surration`).

### CODE REVIEW

**Comment by gabydd** [comment]
In my brief testing this seems to work really well, It think it would be good to test this on a wide range of lsp's. Also it should fix #4131 and maybe some of the related pr's, the issues that have the completion inserted in between text like #4851 probably isn't fixed by this.

**Comment by Taywee** [comment]
In a quick test, this does not seem to fix #4131.  It looks a bit to me like gopls has bad completion in the first place. 

Given this input:

<details><summary>input</summary>
<p>

```
2023-01-10T13:43:53.835 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"textDocument/didChange","params":{"contentChanges":[{"range":{"end":{"character":8,"line":3},"start":{"character":8,"line":3}},"text":"["}],"textDocument":{"uri":"file:///tmp/tmp.APmRKN78A3/hxtest.go","version":14}}}
2023-01-10T13:43:54.283 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"textDocument/didChange","params":{"contentChanges":[{"range":{"end":{"character":9,"line":3},"start":{"character":9,"line":3}},"text":"s"}],"textDocument":{"uri":"file:///tmp/tmp.APmRKN78A3/hxtest.go","version":15}}}
2023-01-10T13:43:54.459 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"textDocument/didChange","params":{"contentChanges":[{"range":{"end":{"character":10,"line":3},"start":{"character":10,"line":3}},"text":"t"}],"textDocument":{"uri":"file:///tmp/tmp.APmRKN78A3/hxtest.go","version":16}}}
2023-01-10T13:43:54.471 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"textDocument/completion","params":{"position":{"character":11,"line":3},"textDocument":{"uri":"file:///tmp/tmp.APmRKN78A3/hxtest.go"}},"id":6}
```
</p>
</details>

The response contains this:

<details><summary>response</summary>
<p>

```json
      {
        "label": "string",
        "labelDetails": {},
        "kind": 7,
        "sortText": "00029",
        "filterText": "string",
        "insertTextFormat": 1,
        "textEdit": {
          "range": {
            "start": {
              "line": 3,
              "character": 11
            },
            "end": {
              "line": 3,
              "character": 11
            }
          },
          "newText": "string"
        }
      }
```
</p>
</details>


I'm not sure at the moment how other client LSPs handle it.  I'll give it a shot with nvim real quick.

**Comment by Taywee** [comment]
In nvim, the gopls suggested response is:

```lua
{ filterText = "string", insertTextFormat = 2, kind = 7, label = "string", labelDetails = vim.empty_dict(), sortText = "00029", textEdit = { newText = "string", range = { ["end"] = { character = 11, line = 3 }, start = { character = 11, line = 3 } } } }
```

So nvim handles the textEdit differently, too.  The gopls issue seems unrelated to what this PR addresses.

**Review by pascalkuthe — APPROVED** [review]
I looked at how nvim-cmp handles this and the LSP spec again and I am pretty sure this is correct. To quote the spec:

> Completion item provides an insertText / label without a text edit: in the model the client should filter against what the user has already typed using the word boundary rules of the language (e.g. resolving the word under the cursor position). The reason for this mode is that it makes it extremely easy for a server to implement a basic completion list and get it filtered on the client.

`start_offset` is the start of the previous word so everything up to that should be replaced.
 
This will technically generate transactions that are a bit larger then they need to be but I doubt that will realistically ever matter as completions are usually pretty small.

**Comment by pascalkuthe** [comment]
The golps issue is a similar but unrelated issue as they do send an text edit

**Comment by Taywee** [comment]
Thanks for the spec reference.  I'm still curious how nvim is handling the gopls case, because the textEdit range appears to not be correct, and nvim still gets the right result.  According to the next section in the spec, a TextEdit with the same start and endpoints should be a plain insert.   I might be simply missing something, but I guess that's not totally relevant to this PR or issue anyway.

**Comment by pascalkuthe** [comment]
> Thanks for the spec reference. I'm still curious how nvim is handling the gopls case, because the textEdit range appears to not be correct, and nvim still gets the right result. According to the next section in the spec, a TextEdit with the same start and endpoints should be a plain insert. I might be simply missing something, but I guess that's not totally relevant to this PR or issue anyway.

I  have been digging intot his and at this point I am pretty sure its actually an upstream issue with the language server. Clearly sending and empty edit range is supposed to be an insertion and they have been receiving and (sometimes) fixing issues for that too. In particular I found the following commdnt:
https://github.com/golang/go/issues/57384#issuecomment-1359455698

I think the response here is odd. They are quoting the lsp standard and say that what they do.is not standard compliant and that is the reason for the bug but then they go on to state that vscode handels this fine and they don't see a reason it shouldn't work so it's ok? Why even have a standard then? 

We could add some kind of workaround (we should port it from vscode if anybody can find where it's happening) although I am in favor of reporting it upstream instead of.bending.over backwards for non standard compliant severs.


I also looked at nvim-cmp and Laplace (gave up with vscode):
* Laplace always extends the removed text to the full word no matter what the edit range says (not spec compliant)
* nvim-cmp used to accidently do the same but got.patched, after that people complained and somebody posted a weird fix thathaopens to work but it seem like it directly.violates the spec and they just didn't notice because the result is fine most of.the time.regardless. There is also no explanation od how/why this wokrs. During PR review the maintainer asked his/why it works, the author said he doesn't know and it got merged anyway? :D

**Comment by Taywee** [comment]
Wow. So no major LSP client is actually spec compliant on TextEdit completions, including VSCode. That's surprising. I'd rather be correct and complain upstream than violate the spec and paper over the difference.

Based on the linked issue, it does seem like they might have fixed the issue in gopls in the main branch as of a few weeks ago: https://github.com/golang/go/issues/57384

**Comment by pascalkuthe** [comment]
Emacs LSP actually also has an open issue about it. They handle it the same way helix does and there are others too (for nvim alone there are like 4? LSP completions clients or more and some of those gopls issue were filed by nvim people so some of them probably handle this correctly). I didn't bother to look for more implementations tough.

 I do agree that we should stick to the standard.
I hope gpls fixes this upstream. It sounded like that the issue I linked was only about some completions for tests tough. I am not sure if they solved it in general or just for that specific case as that issue you mentioned about the general case was closed as duplicate

**Comment by Taywee** [comment]
Right, I'm mostly just shocked about VSCode. If one implementation should be fully spec compliant, it should be the editor that the standard was created for.

**Comment by archseer** [comment]
This probably also solves https://github.com/helix-editor/helix/pull/1819 since we replace everything from the starting offset onwards

**Comment by pascalkuthe** [comment]
> This probably also solves #1819 since we replace everything from the starting offset onwards

This change only affects LSP servers that do not send a `TextEdit` with their completions and leave it up to the editor what text to replace. Most LSPs do send an edit (according to the spec the option of not sending an exit only exists to allow quick imolemtmstion of basic servers) and are unaffectd by this change. So at least to those servers the problem persists.

**Comment by pascalkuthe** [comment]
I missed something while reviewing this PR: It removes the ability to complete at multiple cursors at once. I included this PR in #5728 as I was already making multiple changes to the same code that would cause conflicts and fixed that problem. I kept the commit from this PR to give credit.

**Comment by archseer** [comment]
Was included in https://github.com/helix-editor/helix/pull/5728

**Pull request closed without merging**
